apps/dotcom/src/components/DocumentName/DocumentName.tsx

```diff
@@ -0,0 +1,295 @@
+}
+
+function getName(state: NameState, document: TLDocument, defaultName: string) {
+	if (state.name) return state.name
+	if (document.name === '') {
+		return defaultName
+	}
+	return document.name
+}
+
+const DocumentNameEditor = track(function DocumentNameEditor({
+	state,
+	setState,
+}: {
+	state: NameState
+	setState: (update: SetStateAction<NameState>) => void
+}) {
+	const inputRef = useRef<HTMLInputElement>(null)
+	const editor = useEditor()
+	const documentSettings = editor.getDocumentSettings()
+	const msg = useTranslation()
+	const defaultDocumentName = msg('document.default-name')
+
+	useEffect(() => {
+		if (state.isEditing && inputRef.current) {
+			inputRef.current.select()
+		}
+	}, [state.isEditing])
+
+	useEffect(() => {
+		const save = () => {
+			if (state.name === null) return
+			const trimmed = state.name.trim()
+			if (!trimmed || trimmed === documentSettings.name.trim()) {
+				if (!state.isEditing) setState((prev) => ({ ...prev, name: null }))
+				return
+			}
+
+			editor.updateDocumentSettings({ name: trimmed })
+		}
+
+		if (state.isEditing === false) {
+			save()
+		}
+	}, [documentSettings.name, editor, state.isEditing, state.name, setState])
+
+	const handleChange = useCallback(
+		(e: ChangeEvent<HTMLInputElement>) => {
+			const value = e.currentTarget.value
+			setState((prev) => ({ ...prev, name: value }))
+		},
+		[setState]
+	)
+
+	const handleKeydownCapture = useCallback(
+		(e: KeyboardEvent) => {
+			if (e.key === 'Enter') {
+				e.preventDefault()
+				// blur triggers save
+				inputRef.current?.blur()
+			} else if (e.key === 'Escape') {
+				e.preventDefault()
+				// revert to original name instantly so that when we blur we don't
+				// trigger a save with the new one
+				setState((prev) => ({ ...prev, name: null }))
+				inputRef.current?.blur()
+			}
+		},
+		[setState]
+	)
+
+	const handleBlur = useCallback(() => {
+		setState((prev) => ({ ...prev, isEditing: false }))
+	}, [setState])
+
+	const name = getName(state, documentSettings, defaultDocumentName)
+
+	return (
+		<div className="tlui-document-name__input__wrapper">
+			{state.isEditing && (
+				<input
+					ref={inputRef}
+					className="tlui-document-name__input"
+					value={state.name ?? documentSettings.name}
+					onChange={handleChange}
+					onBlur={handleBlur}
+					onKeyDownCapture={handleKeydownCapture}
+					placeholder={defaultDocumentName}
+					autoFocus
+				/>
+			)}
+			{state.isEditing ? (
+				<div
+					className="tlui-document-name__text tlui-document-name__text__hidden"
+					// @ts-expect-error
+					inert=""
+					aria-hidden
+				>
+					{addRealSpaceForWhitespace(name) || ' '}
+				</div>
+			) : (
+				<div
+					className="tlui-document-name__text"
+					onDoubleClick={() => setState((prev) => ({ ...prev, isEditing: true }))}
+				>
+					{addRealSpaceForWhitespace(name) || ' '}
+				</div>
+			)}
+		</div>
+	)
+})
+
+function addRealSpaceForWhitespace(string: string) {
+	if (string === '') string = ' '
```

Comment:
`!string` ?
a comment here might be useful too - hard to say what this is trying to address. also, it throws me off that above the logic is `addRealSpaceForWhitespace(name) || ' '` (using a regular space, i think for the `|| ' '`) 🤔

Reply:
well, probably the `|| ' '` isn't necessary anyway ¯\_(ツ)_/¯

Reply:
Here I also just [copied the existing code](https://github.com/tldraw/brivate/pull/2972/files#diff-d825ed9304ac39ce1e768f560737448b1af61616404664d42547ed631cd397fcR304-R307), but happy to add some comment.

Reply:
Looking at this again it seems like the `addRealSpaceForWhitespace(name) || ' '` it was probably meant as a guard so that we wouldn't show `undefined / null` in the ui. With use having a `defaultName` fallback now we can probably remove this.